staging/src/k8s.io/kubectl/pkg/config/v1alpha1/types.go

```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Preferences stores elements of KubeRC configuration file
+type Preferences struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Overrides is used to change the defaults values of flags of commands.
+	// This is especially useful, when user doesn't want to explicitly
+	// set flags each time.
+	// +listType=atomic
+	Overrides []CommandOverride `json:"overrides"`
+
+	// Aliases stores the alias definitions. If the alias name collides with
+	// a built-in command, built-in command always overrides the alias name.
+	// +listType=atomic
+	Aliases []AliasOverride `json:"aliases"`
+}
+
+// AliasOverride stores the alias definitions.
+// It is applied in a pre-defined order which is
+// kubectl [ALIAS NAME] expands to kubectl [COMMAND] [USER_FLAGS] [USER_EXPLICIT_ARGUMENTS] [USER_KUBERC_ARGUMENTS]
```

Comment:
> We need to make this documentation explain exactly how expansion will happen, the ordering and precedence.

Tried to update the documentation to better reflect the behavior. Please let me know what you think.

> where do RC alias flags go in this ordering? I would intuitively expect them to go before the explicit user flags (and be replaced if the user specifies the same flag, as discussed below)

With respect to flags, there is no clear ordering but in any case, it aligns with your expectation.

> describe how explicit user flags override RC alias flags

Updated documentation

>  why are user explicit arguments before RC alias arguments? I intuitively expected the other order, but maybe there's a good reason for this ordering?

Tried to explain the reasoning in https://github.com/kubernetes/kubernetes/pull/125230#discussion_r1798932923

> if I define an alias AND an override for the command the alias points to, is the flag override honored? I think it should... 

That would lead to high cognitive load from user's point of view as it includes 3 layers to memorize and spend an effort for the expected behavior. That's why, for the sake of simplicity, there is a clear separation between aliasing and overriding (at least in alpha, unless users demand this functionality). Aliases can't be overridden and overridden flags don't affect aliases. In my opinion, this would also hurt the readability of kuberc. I have to navigate through the file to ensure that flag is not set in another layer. With this version, you are only looking at the definition of alias or override.

> if so, that means we have 3 things to layer... RC alias flags, RC override flags, and explicit user flags. We definitely need tests for this behavior exercising combinations of present/absent at each layer (user, rc alias, rc override).

Fully agree. If we add this, this will require a lot of unit tests.

Reply:
> If we add this, this will require a lot of unit tests.

Even if we don't add it, we still need unit tests demonstrating that overrides do not take effect for aliased commands :)

Reply:
Sounds reasonable and thank you for review :). Added 2 test cases for this.